operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeployment.java

```diff
@@ -92,12 +92,23 @@ public Optional<HasMetadata> getReconciledResource() {
         }
         else {
             Log.info("Existing Deployment found, handling migration");
-
+            
+            // in case existingDeployment is coming from cache, make absolutely sure it needs deleted
+            if (!hasExpectedMatchLabels(existingDeployment) && !hasExpectedMatchLabels(client.resource(existingDeployment).get())) {
```

Comment:
> The reconciliation loop finishes in a desired state. 

Ideally we won't have a blocking delete in the reconciliation loop.  In this case the deletion of the statefulset should be basically immediate - the pods shouldn't block it - but I was still hesitent to add that in.  I think kubectl scales statefulsets to 0 first probably for better behavior in a resource constrained environment, rather than relying on gc to clean up the pods.  But that didn't seem necessary either.

> That means that status fields will indicate KC as ready whereas in reality the deployment is being terminated. Although this is useful as a side effect to [retrigger the reconciliation loop](https://github.com/keycloak/keycloak/blob/58713a23b2b39f55e97bbadee95cceb7c3d3606c/operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java#L143-L146) (that's necessary to re-create the StS again), I believe we should not indicate it's ready. Or perhaps that's ok since the deletion itself immediately triggers another reconciliation loop.

Correct, based upon the current state the status for that loop may indicate it's ready - however that will be ephemal.  The event of marking the deletion timestamp on the statefulset, its eventual deletion, or any other event will trigger another reconciliation, that will update the status to whatever is observed at that point on the statefulset.

The alternative, which I'll change the pr to, is to just let the reconciliation logic proceed as normal after the delete call rather than returning an empty optional - if the statefulset still exists, we'll get an ephemeral error, if not then things will proceed.

> Do we really not trust the cached version of existing StS? I'm not sure it's really desired to explicitly fetch the StS in every reconciliation loop. 

It won't be fetched everytime.  It's only for a final check prior to deletion. 

> Are there any serious concerns that the cached version is incorrect? If so, I'd say it's an JOSDK issue.

This isn't really a JOSDK issue.  The JOSDK mitigates this in a couple of ways - for one all updates / deletes are locked with respect to the resourceVersion.  So it's quite possible for the api server to throw a conflict error when there's been a change that has not yet been picked up by the informer watch - that is part of the motivation for refining the status handling so that the likely very ephemeral error doesn't seem worse than it is.  We could do that check here instead - if the resourceVersion is set on the existing (which is should be except for some of the tests) then issue a delete locked to that version.  I guess I thought that it would be clearer doing another fetch, but I'll change it do a locked operation instead.

Another mechanism in the JOSDK is that your own changes to the resource are tracked in another cache beside the informer cache, so as long as it's you making the changes you don't have to react to your own events.  In the case of a statefulset it can of course have other modifications from the statefulset controller.

Reply:
A consequence of these changes is that it becomes hard to test the migration behavior with the current integration test logic.  We need to create a keycloak instance and modify the associated statefulset and replace it with one that has the old match labels.  More than likely this would require an additional setup step where the keycloak and statefulset could be created priort to operator running, or the introduction of a pause reconciliation annotation so that we could modify the statefulset later, then remove the annotation and let the operator migrate it.

Reply:
Thanks for the elaborate reply.

To me, a locked delete feels more correct than a fetch every time the reconciliation loop runs. The delete will be performed only once, if ever.

I agree testing it would be a bit harder, but I think it's the way to go.

Reply:
> To me, a locked delete feels more correct than a fetch every time the reconciliation loop runs. The delete will be performed only once, if ever.

To clarify it was only doing the fetch if the existing appeared to be in the old state, so that would only occur a limited number of times during migration.

> I agree testing it would be a bit harder, but I think it's the way to go.

Are you in favor of putting in a pause mechanism as part of this pr?

Reply:
> To clarify it was only doing the fetch if the existing appeared to be in the old state, so that would only occur a limited number of times during migration.

Correct, sorry for the confusion from my side. :) Still, locked version feels more correct IMHO.

> Are you in favor of putting in a pause mechanism as part of this pr?

A silly question. Can't we simply manually create the StS in test (prior creating the Keycloak CR) and create the CR after that for Operator to migrate the StS?